Refactor API code for GitHub backend

Benaiah opened the pull request:
<!--
Thanks for submitting a pull request!

Please make sure you've read and understood our contributing guidelines;
https://github.com/netlify/netlify-cms/blob/master/CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx", where #xxxx is the issue number.

Please provide enough information so that others can review your pull request.
The first three fields are mandatory:
-->

**- Summary**

Refactor of `API.js` in the GitHub backend. General improvements to
code style, removal of mutation, function interdependencies, and race
conditions (see the commit message for 732ede1 for more details on
that). Code modified is generally _much_ shorter.

Other refactors I'm working on for this PR:

- [x] Refactor `updateTree`
- [x] Refactor `editorialWorkflowGit`
- [x] Replace `_` with individual function imports.
- ~Use `Immutable` objects throughout to avoid having to convert to and from them in individual functions.~
- ~Refine request handling (possibly remove `meta` field from requesterrors if a better solution is found).~

EDIT: I'm closing the scope of this so we can get it merged. Further refactors can come in later PRs.

<!--
Explain the **motivation** for making this change.
What existing problem does the pull request solve?
-->

**- Test plan**

Tested manually, and existing tests pass.

<!--
Demonstrate the code is solid.
Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI.
-->

**- Description for the changelog**

- Refactor `API.js` in GitHub backend

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://user-images.githubusercontent.com/1425133/28857194-5ed58db4-76fc-11e7-9eb5-2f0808f13060.png)

tortilaman commented on src/backends/github/API.js:
```diff
@@ -20,70 +21,44 @@ export default class API {
-      "Content-Type": "application/json",
+    return {
       ...headers,
+      ...(this.token ? { Authorization: `token ${ this.token }` } : {}),
```
is there a reason to include the empty `{}` when there's no token as opposed to using
`...(this.token && { Authorization: `token ${ this.token}` }),`? I would think unless something is looking for data based on array position, removing an empty object would be preferable.

Benaiah replied:
I don't think this will work, since `...(undefined)` and `...(false)` both throw a `SymbolError`.

tortilaman replied:
Ah, ok.

erquhart replied:
Nitpick/point of discussion: I honestly think this is less readable. Prettier and more succinct, but less readable. Thoughts?

erquhart replied:
Actually I take it back, this whole thing is a massive improvement. Going to abstain from nitpicking in favor of faster :ship:ping.

tortilaman commented on src/backends/github/API.js:
```diff
@@ -20,70 +21,44 @@ export default class API {
-    })
+    return this.request('/user/repos').then(repos =>
+      Object.keys(repos).some(
+        key => (repos[key].full_name === this.repo) && repos[key].permissions.push
```
Ooh. Wasn't aware of `some`. Very useful stuff.

tortilaman commented on src/backends/github/API.js:
```diff
@@ -20,70 +21,44 @@ export default class API {
       const contentType = response.headers.get("Content-Type");
       if (contentType && contentType.match(/json/)) {
-        return this.parseJsonResponse(response);
+        return Promise.all([response, response.json()]);
```
Awesome moving to promises!

Benaiah commented:
Turns out I'd broken `storeMetadata` - it's now fixed. I've also added a big refactor of the `updateTree` function (@tortilaman, if you could review that change as well I'd appreciate it).

tech4him1 commented:
Looks like Travis tests are failing on 8872d18:
> github API › should create PR with correct base branch name when publishing with editorial workflow

Benaiah commented:
@tech4him1 thanks - it was working on my machine, so I foolishly assumed it was being polyfilled. Replacing with `entries` from lodash.

tortilaman commented on src/backends/github/API.js:
```diff
@@ -507,42 +458,35 @@ export default class API {
+        ];
+
+        const newItems = Object.entries(fileTree).filter(([filename]) => !added[filename]);
+        const [newFiles, newDirs] = partition(newItems, ([, file]) => this.isFile(file));
```
Not following the `([, file[)` part of this.

Benaiah replied:
@tortilaman it's destructuring the argument and ignoring the first element. For example:

```
// these are equivalent

const f1 = ([ x, y ]) => y > 0
const f2 = ([, y]) => y > 0

f1(0, 1) // true
f2(0, 1) // true
f1(1, 0) // false
f2(1, 0) // false
```

It allows us to prevent creating a name for something we never use later on. If it's obscure enough it's confusing to most people, I can change it back.

Benaiah commented:
Got another big function refactored - `editorialWorkflowGit`.

Benaiah commented:
Fixed a bug in my `updateTree` refactor which I uncovered while writing my upcoming GitHub API tests.

kalinchernev commented:
I was about to create an issue about several issues related to the API calls to github, but found this pull request. I didn't see any modifications in the tests, so it wasn't very easy to figure if these will be related to any changes already fixed.

I suppose many fixes might be already in via the [request](https://github.com/netlify/netlify-cms/pull/508/files#diff-d3cff7a0c66d0acce37277791446d509R48-R66) refactoring, anyway please test these.

1) Collection folder setting can break the API request silently

For example, I had a config like this:

```yml
collections:
    folder: "content/sample-posts/"
```

The console yields an error because the following request fails `api.github.com/repos/kalinchernev/gatsby-jammin/contents/content/sample-posts//15-08-2017/test.md?ts=1502817755076&ref=master` - it has double slash.

This configuration fixes this issue

```yml
collections:
    folder: "content/sample-posts"
```
Possible improvement here could be that at least there is a simple component in the body that gives feedback to the user there is an internal error. And of course having some checks for this config entry to ensure correct request to the github API.

2) The list of posts (or whatever collection type) does not show existing items

I'm using Gatsby, and structure is 

```
.
├── 2017-08-13-test.md
└── sample-posts
    ├── 01-01-2017
    │   └── index.md
    ├── 01-02-2017
    │   └── index.md
    ├── 01-03-2017
    │   └── index.md
    ├── 01-07-2017-Big_Sample_Post
    │   └── index.md
    ├── 02-03-2017
    │   └── index.md
    ├── 02-04-2017
    │   └── index.md
    ├── 15-08-2017
    │   └── test2.md
    ├── 17-03-2017
    │   └── index.md
    └── 23-03-2017
        └── index.md
```

None of these are displayed in the list of content, even when content has been created via the workflow.

Repo is [here](https://github.com/kalinchernev/gatsby-jammin) if that might help.

Benaiah commented:
@kalinchernev the first issue is definitely related to how we construct requests - we should be checking to avoid doubling slashes at the beginning and end of user-supplied strings. The other problem appears to be due to the fact that each of your entries is in its own subfolder - this isn't currently supported by the CMS. You're welcome to make an issue regarding that if you want to discuss it further, but it's a bigger concern than those addressed by this PR, which is mostly focused on code cleanup.

erquhart commented:
@kalinchernev see https://github.com/netlify/netlify-cms/issues/513 for the related improvement.

kalinchernev commented:
@erquhart thanks, I didn't open an issue as suggested, as I managed by flattening as discussed in the existing issue :)

erquhart commented on src/backends/github/API.js:
```diff
@@ -20,70 +21,47 @@ export default class API {
-    return baseHeader;
-  }
-
-  parseJsonResponse(response) {
```
Nice fat trimmin'

erquhart replied:
Update: the Git Gateway backend is now using this as well - maybe worth keeping and refactoring.

erquhart commented on src/backends/github/API.js:
```diff
@@ -229,50 +208,26 @@ export default class API {
+    return files
+      .map(file => ({ ...file, file: true }))
+      .reduce((tree, file) => tree.setIn(file.path.split("/"), file), Map())
+      .toJS();
```
😍

begincalendar replied:
Does it matter that the original version did not include nested keys that were empty and the new one does?
I.e. `.filter(part => part)` appears to have been dropped.

erquhart commented on src/backends/github/API.js:
```diff
@@ -229,50 +208,26 @@ export default class API {
-    });
+    return Promise.all([fileTreePromise, this.getBranch()])
+      .then(([fileTree, branchData]) => this.updateTree(branchData.commit.sha, "/", fileTree))
+      .then(changeTree => this.commit(options.commitMessage, changeTree));
```
No need for patch?

tech4him1 replied:
Wondering that too, was it moved to a different location?

erquhart replied:
I'm assuming it was superfluous, but wanted to confirm.

Benaiah replied:
I don't remember removing this intentionally, so it was likely an oversight. Thanks for the catch!

erquhart commented on src/backends/github/API.js:
```diff
@@ -282,94 +237,70 @@ export default class API {
+      timeStamp: new Date().toISOString(),
+    };
+
+    return (unpublished
```
A lot of this really should be broken down and notated for clarity at some point in the future.

Benaiah replied:
Definitely agreed.

tech4him1 commented on src/backends/github/API.js:
```diff
@@ -507,44 +437,37 @@ export default class API {
+    return obj.file;
+  }
+
   updateTree(sha, path, fileTree) {
```
I'm currently having a problem with `updateTree` where sometimes it will request a tree update, but with `tree` set to `[{}]` (empty). I'm leaving this note here as a block until I can figure out what the error is.

tech4him1 commented on src/backends/github/API.js:
```diff
@@ -507,44 +437,37 @@ export default class API {
-          .then(updates => this.request(`${ this.repoURL }/git/trees`, {
+          .then(resolvedUpdates => this.request(`${ this.repoURL }/git/trees`, {
             method: "POST",
             body: JSON.stringify({ base_tree: sha, tree: updates }),
```
Shouldn't the `tree` be `resolvedUpdates` here?

I'm was having a problem with `updateTree` where sometimes it will request a tree update, but with tree set to `[{}]` (empty).

Benaiah replied:
Great catch, thanks.

tech4him1 replied:
@Benaiah Just wondering, are you guys planning on getting the tests done before you merge this?

erquhart replied:
This still needs to be fixed.

erquhart commented:
@Benaiah we can look at merging once @tech4him1's comments are addressed.

dopry commented:
@Benaiah, hey I tried a rebase. The only conflict I ran into was 

* isCollaborator has been superceded by hasWriteAccess

I have pushed my rebase to https://github.com/dopry/netlify-cms/tree/gh-api-refactors if you want to pull it and update the PR.

Benaiah commented:
@dopry great, thanks!

alexanderankin commented:
any updates on this? thanks to all in this thread!

erquhart commented:
@alexanderankin we de-prioritized this a bit because we're headed for 1.0, and refactoring the GH backend carries some unwelcome risk.

alexanderankin commented:
ok sometime in the future I am planning on spending time figuring out how to augment this tool to support gitlab so im taking my sweet time aquianting myself with all the requisite stuff. thanks for the response, I have a lot catching up to do.

tech4him1 commented:
@alexanderankin We do have a GitLab PR at #517 that is working, we just de-prioritized that one as well for a bit.

tech4him1 commented on src/backends/github/API.js:
```diff
@@ -20,70 +21,47 @@ export default class API {
     })
-    .catch((error) => {
-      throw new APIError(error.message, responseStatus, 'GitHub');
+    .catch(err => [err, null])
```
@Benaiah [You mentioned that](https://github.com/netlify/netlify-cms/pull/517#discussion_r135940695):
> This should be `.catch(err => Promise.reject([err, null))` so the other `.catch` block catches and destructures it properly. This is my bad - it's the same way in my GH API refactor PR.

Benaiah replied:
That's correct, thanks!

begincalendar commented on src/backends/github/API.js:
```diff
@@ -20,70 +21,47 @@ export default class API {
+          ? Object.entries(options.params).map(
+            ([key, val]) => `${ key }=${ encodeURIComponent(val) }`)
+          : [];
+    return `${ this.api_root }${ path }?${ [cacheBuster, ...encodedParams].join("&") }`;
```
Can we make use of modern API here (to avoid re-inventing the wheel)?
Specifically:

- [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) for the query string
- [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL) for the merging of parts (root, path and query string)

tech4him1 replied:
Just a note: `URLSearchParams` isn't supported in MS Edge <17. Otherwise, I think this is a great idea.

erquhart commented:
Closing as stale, leaving branch deletion for @Benaiah.
